### PR TITLE
proc refactoring: make stack, disassemble and eval independent of proc.Process

### PR DESCRIPTION
**Pull request by aarzilli**
Second step on the way to #782

### CODE REVIEW

File: pkg/proc/disasm.go [hunk 1]
```diff
@@ -16,14 +16,27 @@ const (
 	IntelFlavour
 )
 
+func (dbp *Process) Disassemble(g *G, startPC, endPC uint64) ([]AsmInstruction, error) {
```
Comment (derekparker) [inline pkg/proc/disasm.go]: Needs doc comment.

File: pkg/proc/disasm.go [hunk 2]
```diff
@@ -16,14 +16,27 @@ const (
...
+
 // Disassemble disassembles target memory between startPC and endPC
 // If currentGoroutine is set and thread is stopped at a CALL instruction Disassemble will evaluate the argument of the CALL instruction using the thread's registers
 // Be aware that the Bytes field of each returned instruction is a slice of a larger array of size endPC - startPC
-func (thread *Thread) Disassemble(startPC, endPC uint64, currentGoroutine bool) ([]AsmInstruction, error) {
-	if thread.dbp.exited {
-		return nil, &ProcessExitedError{}
-	}
-	mem, err := thread.readMemory(uintptr(startPC), int(endPC-startPC))
+func Disassemble(memrw memoryReadWriter, regs Registers, breakpoints map[uint64]*Breakpoint, bi *BinaryInfo, startPC, endPC uint64) ([]AsmInstruction, error) {
```
Comment (derekparker) [inline pkg/proc/disasm.go]: Seems like a cleaner signature would be to simply pass in a `[]byte` representing the read memory, and let this function focus just on the actual parsing, disassembling.

File: pkg/proc/disasm_amd64.go [hunk 3]
```diff
@@ -63,7 +63,7 @@ func (inst *AsmInstruction) IsCall() bool {
 	return inst.Inst.Op == x86asm.CALL || inst.Inst.Op == x86asm.LCALL
 }
 
-func (thread *Thread) resolveCallArg(inst *ArchInst, currentGoroutine bool, regs Registers) *Location {
+func resolveCallArg(inst *ArchInst, currentGoroutine bool, regs Registers, mem memoryReadWriter, bininfo *BinaryInfo) *Location {
```
Comment (derekparker) [inline pkg/proc/disasm_amd64.go]: Big fan of taking these ugly unexported methods off of their callers! :+1:

File: pkg/proc/disasm_amd64.go [hunk 4]
```diff
@@ -143,10 +139,14 @@ func init() {
 	}
 }
 
+func (dbp *Process) FirstPCAfterPrologue(fn *gosym.Func, sameline bool) (uint64, error) {
```
Comment (derekparker) [inline pkg/proc/disasm_amd64.go]: Needs doc comment.

File: pkg/proc/eval.go [hunk 5]
```diff
@@ -67,7 +67,11 @@ func (scope *EvalScope) evalAST(t ast.Expr) (*Variable, error) {
 		// try to interpret the selector as a package variable
 		if maybePkg, ok := node.X.(*ast.Ident); ok {
 			if maybePkg.Name == "runtime" && node.Sel.Name == "curg" {
-				return scope.Thread.getGVariable()
+				if scope.gvar == nil {
+					return nilVariable, nil
+				} else {
```
Comment (derekparker) [inline pkg/proc/eval.go]: Don't need this `else` clause since we're returning.

File: pkg/proc/moduledata.go [hunk 6]
```diff
@@ -11,9 +11,9 @@ type moduleData struct {
 	typemapVar    *Variable
 }
 
-func (bi *BinaryInfo) loadModuleData(thread *Thread) (err error) {
+func (bi *BinaryInfo) loadModuleData(mem memoryReadWriter) (err error) {
```
Comment (derekparker) [inline pkg/proc/moduledata.go]: Would kind of like to take these unexported methods and turn them into just functions that accept `bi *BinaryInfo` ... just seems cleaner to me.

File: pkg/proc/proc.go [hunk 7]
```diff
@@ -179,36 +179,44 @@ func (dbp *Process) LoadInformation(path string) error {
 	return nil
 }
 
+func (dbp *Process) FindFileLocation(fileName string, lineno int) (uint64, error) {
```
Comment (derekparker) [inline pkg/proc/proc.go]: Needs doc comment.

File: pkg/proc/proc.go [hunk 8]
```diff
@@ -179,36 +179,44 @@ func (dbp *Process) LoadInformation(path string) error {
...
 		return 0, err
 	}
 	if fn.Entry == pc {
-		pc, _ = dbp.FirstPCAfterPrologue(fn, true)
+		pc, _ = FirstPCAfterPrologue(mem, breakpoints, bi, fn, true)
 	}
 	return pc, nil
 }
 
+func (dbp *Process) FindFunctionLocation(funcName string, firstLine bool, lineOffset int) (uint64, error) {
```
Comment (derekparker) [inline pkg/proc/proc.go]: Needs doc comment.

File: pkg/proc/variables.go [hunk 9]
```diff
@@ -132,15 +132,16 @@ type G struct {
...
-	dbp      *Process
 }
 
 // EvalScope is the scope for variable evaluation. Contains the thread,
 // current location (PC), and canonical frame address.
 type EvalScope struct {
-	Thread *Thread
-	PC     uint64
-	CFA    int64
+	PC   uint64
```
Comment (derekparker) [inline pkg/proc/variables.go]: nit: Can you add comments to exported struct members.

File: pkg/proc/disasm.go (hunk 2 above)
Comment (aarzilli) [inline pkg/proc/disasm.go]: For the most barebones functionality, yes. But we are also marking breakpoint and PC positions and resolving the CALL argument whenever possible. In the future I would also like to resolve SP relative addresses to local variable names.

File: pkg/proc/disasm_amd64.go (hunk 4 above)
Comment (aarzilli) [inline pkg/proc/disasm_amd64.go]: Both for this and Disassemble I've added a doc comment to target.Interface.

**Comment by aarzilli** [comment]
Done.

**Pull request merged**